src/libostree/ostree-bloom.c

```diff
@@ -0,0 +1,556 @@
+ */
+gboolean
+ostree_bloom_maybe_contains (OstreeBloom   *bloom,
+                             gconstpointer  element)
+{
+  guint8 i;
+
+  g_return_val_if_fail (bloom != NULL, TRUE);
+  g_return_val_if_fail (bloom->ref_count >= 1, TRUE);
+
+  for (i = 0; i < bloom->k; i++)
+    {
+      gsize idx;
+
+      idx = bloom->hash_func (element, i);
+
+      if (!ostree_bloom_get_bit (bloom, idx % (bloom->n_bytes * 8)))
+        return FALSE;  /* definitely not in the set */
+    }
+
+  return TRUE;  /* possibly in the set */
+}
+
+/**
+ * ostree_bloom_seal:
+ * @bloom: an #OstreeBloom
+ *
+ * TODO; OK to call multiple times
+ *
+ * Returns: (transfer full): a #GBytes containing the immutable filter data
+ * Since: 2017.7
+ */
+GBytes *
+ostree_bloom_seal (OstreeBloom *bloom)
+{
+  g_return_val_if_fail (bloom != NULL, NULL);
+  g_return_val_if_fail (bloom->ref_count >= 1, NULL);
+
+  if (bloom->is_mutable)
+    {
+      bloom->is_mutable = FALSE;
+      bloom->immutable_bytes = g_bytes_new_take (g_steal_pointer (&bloom->mutable_bytes), bloom->n_bytes);
+    }
+
+  return g_bytes_ref (bloom->immutable_bytes);
+}
+
+/**
+ * ostree_bloom_add_element:
+ * @bloom: an #OstreeBloom
+ * @element: (nullable): element to add to the filter
+ *
+ * TODO; nullability depends on hash function
+ *
+ * Since: 2017.7
+ */
+void
+ostree_bloom_add_element (OstreeBloom   *bloom,
+                          gconstpointer  element)
+{
+  guint8 i;
+
+  g_return_if_fail (bloom != NULL);
+  g_return_if_fail (bloom->ref_count >= 1);
+  g_return_if_fail (bloom->is_mutable);
+
+  for (i = 0; i < bloom->k; i++)
+    {
+      gsize idx = bloom->hash_func (element, i);
+      ostree_bloom_set_bit (bloom, idx % (bloom->n_bytes * 8));
+    }
+}
+
+/**
+ * ostree_bloom_get_size:
+ * @bloom: an #OstreeBloom
+ *
+ * Get the size of the #OstreeBloom filter, in bytes, as configured at
+ * construction time.
+ *
+ * Returns: the bloom filter’s size in bytes, guaranteed to be >0
+ * Since: 2017.7
+ */
+gsize
+ostree_bloom_get_size (OstreeBloom *bloom)
+{
+  g_return_val_if_fail (bloom != NULL, 0);
+
+  return bloom->n_bytes;
+}
+
+/**
+ * ostree_bloom_get_k:
+ * @bloom: an #OstreeBloom
+ *
+ * Get the `k` value from the #OstreeBloom filter, as configured at
+ * construction time.
+ *
+ * Returns: the bloom filter’s `k` value, guaranteed to be >0
+ * Since: 2017.7
+ */
+guint8
+ostree_bloom_get_k (OstreeBloom *bloom)
+{
+  g_return_val_if_fail (bloom != NULL, 0);
+
+  return bloom->k;
+}
+
+/**
+ * ostree_bloom_get_hash_func:
+ * @bloom: an #OstreeBloom
+ *
+ * Get the #OstreeBloomHashFunc from the #OstreeBloom filter, as configured at
+ * construction time.
+ *
+ * Returns: the bloom filter’s universal hash function
+ * Since: 2017.7
+ */
+OstreeBloomHashFunc
+ostree_bloom_get_hash_func (OstreeBloom *bloom)
+{
+  g_return_val_if_fail (bloom != NULL, NULL);
+
+  return bloom->hash_func;
+}
+
+/* SipHash code adapted from https://github.com/veorq/SipHash/blob/master/siphash.c */
```

Comment:
OK. If you would prefer to go with systemd, that’s not a problem — but since the hash implementation is API (needs to be the same on the client and server sides of a peer-to-peer resolve operation), it needs to change now or never.